staging/src/k8s.io/apimachinery/pkg/util/waitgroup/waitgroup.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package waitgroup
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// A WaitGroup must not be copied after first use.
+type WaitGroup struct {
+	m       sync.Mutex
+	counter int32
+	ch      chan struct{}
+}
+
+// Add adds delta, which may be negative, to the WaitGroup counter.
+// If the counter becomes zero, all goroutines blocked on Wait are released.
+// If the counter goes negative, Add panics.
+func (wg *WaitGroup) Add(delta int) {
+	v := atomic.AddInt32(&wg.counter, int32(delta))
+	// Panic if negative
+	if v < 0 {
+		panic("sync: negative WaitGroup counter")
+	} else if v == 0 {
+		wg.m.Lock()
```

Comment:
better defer the Unlock. Makes it easier to reason about in case of panics.

Reply:
Sure